src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -402,29 +402,34 @@ bool ObjectMonitor::enter(JavaThread* current) {
     }
 
     OSThreadContendState osts(current->osthread());
-    ThreadBlockInVM tbivm(current);
 
-    // TODO-FIXME: change the following for(;;) loop to straight-line code.
-    for (;;) {
-      current->set_suspend_equivalent();
-      // cleared by handle_special_suspend_equivalent_condition()
-      // or java_suspend_self()
+    assert(current->thread_state() == _thread_in_vm, "invariant");
 
+    current->frame_anchor()->make_walkable(current);
+    OrderAccess::storestore();
+    for (;;) {
+      current->set_thread_state(_thread_blocked);
       EnterI(current);
-
-      if (!current->handle_special_suspend_equivalent_condition()) break;
-
-      // We have acquired the contended monitor, but while we were
-      // waiting another thread suspended us. We don't want to enter
-      // the monitor while suspended because that would surprise the
-      // thread that suspended us.
-      //
-      _recursions = 0;
-      _succ = NULL;
-      exit(false, current);
-
-      current->java_suspend_self();
+      current->set_thread_state_fence(_thread_blocked_trans);
+      if (SafepointMechanism::should_process(current) &&
+        current->suspend_request_pending()) {
+        // We have acquired the contended monitor, but while we were
+        // waiting another thread suspended us. We don't want to enter
+        // the monitor while suspended because that would surprise the
+        // thread that suspended us.
+        _recursions = 0;
+        _succ = NULL;
+        exit(false, current);
```

Comment:
You'll have a conflict with @pchilano recent fix.

Reply:
Fixed with a merge with master.